File: Python/TrappingRainWater_42.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        l=0
        r=len(height)-1
        l_max=0
        r_max=0
        total=0

        while l <= r:
            if height[l] <= height[r]:
                if l_max <= height[l]:
                    l_max=height[l]
                else:
                    total+=l_max-height[l]
                l+=1
            else:
                if r_max > height[r]:
                    total+=r_max-height[r]
                else:
                    r_max=height[r]
                r-=1
        return total


s=Solution()

heights=[0,1,0,2,1,0,1,3,2,1,2,1]

print("Water it can trap: ",s.trap(heights))

File: LeetCode/Java/TrappingRainWater_42.java
public class TrappingRainWater_42 {
    public int trap(int [] height){
        int l=0,r=height.length-1,l_max=0,r_max=0,total=0;

        while( l <= r){
            if( height[l] <= height[r]){
                if (height[l] >= l_max){
                    l_max=height[l];
                }else{
                    total+=l_max - height[l];
                }
                l++;
            }
            else{
                if (height[r] >= r_max){
                    r_max=height[r];
                }else{
                    total+=r_max-height[r];
                }
            r--;
            }
        }
        return total;
    }
    public static void main(String[] args) {
        int[] height={0,1,0,2,1,0,1,3,2,1,2,1};
        TrappingRainWater_42 obj = new TrappingRainWater_42();
        System.out.println("Water it can trap: "+ obj.trap(height));

    }
}
